src/audio_core/hle/wmf_decoder_utils.cpp

```diff
@@ -0,0 +1,366 @@
+        ReportError("Unable to create an empty MediaType", hr);
+        return -1;
+    }
+
+    // basic definition
+    t->SetGUID(MF_MT_MAJOR_TYPE, MFMediaType_Audio);
+    t->SetGUID(MF_MT_SUBTYPE, audio_format);
+
+    // see https://docs.microsoft.com/en-us/windows/desktop/medfound/aac-decoder#example-media-types
+    // and https://docs.microsoft.com/zh-cn/windows/desktop/api/mmreg/ns-mmreg-heaacwaveinfo_tag
+    // for the meaning of the byte array below
+
+    // for integrate into a larger project, it is recommended to wrap the parameters into a struct
+    // and pass that struct into the function
+    // const UINT8 aac_data[] = { 0x01, 0x00, 0xfe, 00, 00, 00, 00, 00, 00, 00, 00, 00, 0x11, 0x90
+    // }; 0: raw aac 1: adts 2: adif 3: latm/laos
+    t->SetUINT32(MF_MT_AAC_PAYLOAD_TYPE, 1);
+    t->SetUINT32(MF_MT_AUDIO_NUM_CHANNELS, adts.channels);
+    t->SetUINT32(MF_MT_AUDIO_SAMPLES_PER_SECOND, adts.samplerate);
+    // 0xfe = 254 = "unspecified"
+    t->SetUINT32(MF_MT_AAC_AUDIO_PROFILE_LEVEL_INDICATION, 254);
+    t->SetUINT32(MF_MT_AUDIO_BLOCK_ALIGNMENT, 1);
+    t->SetBlob(MF_MT_USER_DATA, user_data, user_data_len);
+    hr = transform->SetInputType(in_stream_id, t, 0);
+    if (FAILED(hr)) {
+        ReportError("failed to select input types for MFT", hr);
+        return -1;
+    }
+
+    return 0;
+}
+
+int select_output_mediatype(IMFTransform* transform, int out_stream_id, GUID audio_format) {
+    HRESULT hr = S_OK;
+    UINT32 tmp;
+    IMFMediaType* t;
+
+    // If you know what you need and what you are doing, you can specify the condition instead of
+    // searching but it's better to use search since MFT may or may not support your output
+    // parameters
+    for (DWORD i = 0;; i++) {
+        hr = transform->GetOutputAvailableType(out_stream_id, i, &t);
+        if (hr == MF_E_NO_MORE_TYPES || hr == E_NOTIMPL) {
+            return 0;
+        }
+        if (FAILED(hr)) {
+            ReportError("failed to get output types for MFT", hr);
+            return -1;
+        }
+
+        hr = t->GetUINT32(MF_MT_AUDIO_BITS_PER_SAMPLE, &tmp);
+
+        if (FAILED(hr))
+            continue;
+        // select PCM-16 format
+        if (tmp == 32) {
+            hr = t->SetUINT32(MF_MT_AUDIO_BLOCK_ALIGNMENT, 1);
+            if (FAILED(hr)) {
+                ReportError("failed to set MF_MT_AUDIO_BLOCK_ALIGNMENT for MFT on output stream",
+                            hr);
+                return -1;
+            }
+            hr = transform->SetOutputType(out_stream_id, t, 0);
+            if (FAILED(hr)) {
+                ReportError("failed to select output types for MFT", hr);
+                return -1;
+            }
+            return 0;
+        } else {
+            continue;
+        }
+
+        return -1;
+    }
+
+    ReportError("MFT: Unable to find preferred output format", E_NOTIMPL);
+    return -1;
+}
+
+int detect_mediatype(char* buffer, size_t len, ADTSData* output, char** aac_tag) {
```

Comment:
`bool DetectMediatype` also all the pointer could just be references, or maybe avoid this function completely bu juts using a ADTS constructor